### PR TITLE
[Kernel][Core] Add AWQ support to the Marlin kernel 

### PR DESCRIPTION
**Pull request by alexm-redhat**
This PR adds end-to-end support for AWQ quantization inside the Marlin kernel.

1. Modifies the existing gptq_marlin.cu to support zero-points
2. Adds a new AWQ on-the-fly repack kernel that converts from AWQ format to Marlin format
3. Add a new linear layer: awq_marlin.py

Here are initial performance results of awq_marlin (this PR) vs awq (on the vllm main) for Llama3-70b AWQ model on 2xA100 GPUs and Llama3-8B AWQ on 1xA100 GPU.

_Llama3-70B AWQ on 2xA100 GPUs with prompt = 512 and decode = 256_
image

_Llama3-7B AWQ on 1xA100 GPUs with prompt = 1024 and decode = 512_
image

TODOs (may be done after this PR lands):
1. Add an end-to-end correctness test for AWQ models like we have for GPTQ models for both float16 and bfloat16
2. Anything else?

### CODE REVIEW

**Comment by alexm-neuralmagic** [comment]
@mgoin @robertgshaw2-neuralmagic added bfloat16 support

File: csrc/quantization/gptq_marlin/awq_marlin_repack.cu [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+  cudaDeviceGetAttribute(&max_shared_mem,
+                         cudaDevAttrMaxSharedMemoryPerBlockOptin, dev);
+  TORCH_CHECK(max_shared_mem > 0);
+
+  if (false) {
+  }
+  CALL_IF(4)
+  CALL_IF(4)
+  CALL_IF(8)
+  CALL_IF(8)
```
[the remark this answers is not in the document]
Comment (alexm-neuralmagic) [reply csrc/quantization/gptq_marlin/awq_marlin_repack.cu]: Good catch, removed duplicates

File: csrc/quantization/gptq_marlin/awq_marlin_repack.cu [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+      false, "marlin_repack_from_gptq(..) requires CUDA_ARCH >= 8.0");
+  return torch::empty({1, 1});
+}
+
+#else
+
+namespace marlin {
+
+template <int const num_threads, int const num_bits>
+__global__ void awq_marlin_repack_kernel(
```
[the remark this answers is not in the document]
Comment (alexm-neuralmagic) [reply csrc/quantization/gptq_marlin/awq_marlin_repack.cu:27]: The internal unpacking is different: AWQ packs over columns, while GPTQ over rows, and also AWQ performs the interleaving of groups of 8 (for 4-bit) or groups of 4 (for 8-bit) to be compatible to the de-quantization PTX assembly.

**Comment by casper-hansen** [comment]
As I told @robertgshaw2-neuralmagic on Discord, this kind of speedup warrants a backport to AutoAWQ. 

Does it make sense to natively pack the weights in AutoAWQ for the Marlin format and if so, do you have any reference code for this now that zero points are supported?

**Comment by mgoin** [comment]
@casper-hansen Here are my thoughts: I think it is best to keep the original AutoAWQ checkpoint format as this is what users currently depend on. We had this same question with AutoGPTQ, where it quickly became annoying and confusing to have GPTQ models in the original or Marlin format. 
I consider the checkpoint formats of GPTQ and AWQ to be connected to the quantization of the model at this point, where Marlin is simply a format for runtime execution. This also makes it easier for us in vLLM since we only need to deal with the regular AWQ format, which we want in order to use all the existing checkpoints out there. My stance is it should be up to the inference system to decide how it wants to repack and run the existing checkpoints in the community.

**Review by robertgshaw2-redhat — APPROVED** [review]
Thanks Alex, this looks great.

**Comment by alexm-neuralmagic** [comment]
@yetionyo I was using "casperhansen/llama-3-70b-instruct-awq" from Hugging face with 2xA100 GPUs with NVLink. The prompt was 512 and decode 256, and I was using the benchmark_serving.py script that ships with vllm package to measure TTFT and TPOT. It may be that you have bottleneck somewhere else that hides the benefit of AWQ marlin. In general, on smaller GPUs with limited bandwidth, the improvement of AWQ marlin should be even better than A100.

**Comment by robertgshaw2-neuralmagic** [comment]
> @alexm-neuralmagic Very curious about the config of the test on the llama3-70b?
> 
> I test awq-gemm ckpt by forcing the --quantization to awq, and test the awq-marlin by not setting the quantization method (and the log prints 'The model is convertible to awq_marlin during runtime. Using awq_marlin kernel.'), but only see a around 5% improvement with awq-marlin.
> 
> My test env is 4x4090. Prompt input length is 1024, and output 128 tokens. I have tested on 1~5 concurrent requests, and the improvement are around 5%.

Marlin provides speedup in the batch ~8-64 regime. 

When you say concurrent requests --> is this 1-5 queries per second for an extended period of time?

In alex's setup, even 1 QPS results in easily batch 16, because each request takes ~10seconds ... Marlin targets the online inference serving case. So you won't really see a speedup unless you are putting the server under load

**Comment by alexm-neuralmagic** [comment]
If you specify "-q awq", then this will force run the old awq kernel. To run the new one, either do not specify anything or specify "-q marlin". vLLM should print "The model is convertible to awq_marlin during runtime. Using awq_marlin kernel.".

**Comment by robertgshaw2-neuralmagic** [comment]
We will fix the issue, but if you just drop —quantization awq_marlin it will use the marlin kernel by default

**Comment by robertgshaw2-neuralmagic** [comment]
@alexm-neuralmagic can you post a PR

**Comment by alexm-neuralmagic** [comment]
@casper-hansen @robertgshaw2-neuralmagic here is a PR to fix flags: https://github.com/vllm-project/vllm/pull/6745, should land soon.

**Comment by alexm-neuralmagic** [comment]
@yetionyo would be good to have some reproduction instructions, then we can take a look.

**Comment by kulievvitaly** [comment]
Test is finished after bugfixes. Tested with docker on my 4x3090ti. GPU connected to server motherboard(epyc 7002) with pcie x16 gen4. Ubuntu 24.04, nvidia drivers 555.

`sudo docker run --shm-size=32g --log-opt max-size=10m --log-opt max-file=1 --rm -it --gpus '"device=0,1,2,3"' -p 9000:8000 --mount type=bind,source=/home/me/.cache,target=/root/.cache vllm/vllm-openai:v0.5.3.post1 --model hugging-quants/Meta-Llama-3.1-70B-Instruct-AWQ-INT4 --tensor-parallel-size 4 --dtype half --max_model_len 64000`

My own benchmark:
https://github.com/kulievvitaly/rus_gpt_demo/blob/master/benchmark.py

extract prompt(about 500 in tokens and 5 out tokens)
genarate prompt(about 10 in tokens and 500 out tokens)

great job, thank you!

1, 10, 50 parallel workers.

Results:
Awq_marlin kernel is faster! Extract prompt faster **+8-15%**. Generate prompt faster **+13-23%**. Full results on github.

**Comment by kulievvitaly** [comment]
I have make the same benchmark on A100.
extract **+67-137%**
generate **+44-138%**

The speed increase on the A100 is comparable to the first message in the thread.

**Comment by linpan** [comment]
—quantization awq_marlin  not work.

**Comment by alexm-neuralmagic** [comment]
@linpan should work on latest main.

**Comment by zhyncs** [comment]
Hi @alexm-neuralmagic Nice work! Alex. The performance of Marlin AWQ amazed me, excellent work.

**Comment by alexm-neuralmagic** [comment]
@zhyncs thanks for doing these benchmarks! I would also expect AWQ to be inherently faster than GPTQ because AWQ has no activation order, especially for multi-gpu runs.

**Comment by zhyncs** [comment]
Hi @alexm-neuralmagic Thank you for your reply. I conducted an eval of gsm8k using lm_eval on Llama 3 8B Instruct and the AWQ model, and found that there was a significant decrease in accuracy. Is this expected? My replication steps and results are as follows:
```bash
python -m vllm.entrypoints.openai.api_server --model meta-llama/Meta-Llama-3-8B-Instruct
python -m vllm.entrypoints.openai.api_server --model casperhansen/llama-3-8b-instruct-awq

lm_eval --model local-completions --tasks gsm8k --num_fewshot 8 --model_args model=meta-llama/Meta-Llama-3-8B-Instruct,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=128,max_retries=3,tokenized_requests=True
lm_eval --model local-completions --tasks gsm8k --num_fewshot 8 --model_args model=casperhansen/llama-3-8b-instruct-awq,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=128,max_retries=3,tokenized_requests=True
```

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.7506|±  |0.0119|
|     |       |strict-match    |     8|exact_match|↑  |0.7498|±  |0.0119|

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.7043|±  |0.0126|
|     |       |strict-match    |     8|exact_match|↑  |0.7051|±  |0.0126|
```

**Comment by zhyncs** [comment]
Hi @alexm-neuralmagic I tested vLLM 0.5.4 and noticed the accuracy has worsened. Is this expected? Thanks.

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.6998|±  |0.0126|
|     |       |strict-match    |     8|exact_match|↑  |0.6998|±  |0.0126|
```

**Comment by alexm-neuralmagic** [comment]
@zhyncs thanks for checking. Could you please provide reproduction instructions

**Comment by zhyncs** [comment]
> @zhyncs thanks for checking. Could you please provide reproduction instructions

ref https://github.com/vllm-project/vllm/pull/6612#issuecomment-2265735040

**Comment by robertgshaw2-neuralmagic** [comment]
So the drop is from `0.7043` to `0.6998`?

**Comment by zhyncs** [comment]
Yes, the previous version already had accuracy issues, which this version was supposed to fix, but it ended up being even worse.

**Comment by alexm-neuralmagic** [comment]
@zhyncs did you try --quantization awq (to force the original awq kernel)

**Comment by zhyncs** [comment]
This has nothing to do with that, let's just compare it directly to fp16. The current drop in accuracy is unacceptable, it can't be used in online business at all. By the way, LMDeploy is much better than this, you can test it yourself. Thanks.

**Comment by robertgshaw2-neuralmagic** [comment]
The AWQ model is 4 bit quantized, so you should not expect to see the exact same scores between fp16 and int4

What score does LMDeploy achieve for the AWQ model with GSM?

**Comment by zhyncs** [comment]
> expect to see the same scores

Your understanding has a big problem. From beginning to end, there was never any mention or expectation of the same score. I don't know why you have this strange misunderstanding. What is being said here is that the current implementation of AWQ Marlin drops points too severely. It's not that it can't drop points, but at least you must ensure that the accuracy after dropping points is usable. Right now, this situation belongs to an unusable state.

**Comment by robertgshaw2-neuralmagic** [comment]
I get the same accuracy scores when using `awq_marlin` vs using `awq`:

Client launch command:
```bash
lm_eval --model local-completions --tasks gsm8k --num_fewshot 8 --model_args model=casperhansen/llama-3-8b-instruct-awq,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=128,max_retries=3,tokenized_requests=True
```

- `awq`:
launch command (forces using the old awq kernels):
```bash
python -m vllm.entrypoints.openai.api_server --model casperhansen/llama-3-8b-instruct-awq --quantization awq
```

scores:
```bash
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.7005|±  |0.0126|
|     |       |strict-match    |     8|exact_match|↑  |0.6998|±  |0.0126|
```

- `awq_marlin`:

launch command (we use the `awq_marlin` kernels by default:

```bash
python -m vllm.entrypoints.openai.api_server --model casperhansen/llama-3-8b-instruct-awq
``` 

scores:
```bash
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.6998|±  |0.0126|
|     |       |strict-match    |     8|exact_match|↑  |0.6990|±  |0.0126|
```

**Comment by zhyncs** [comment]
I fully understand your comparison method and your results. Perhaps we should focus on how to improve the current AWQ's accuracy, what do you think? @robertgshaw2-neuralmagic @alexm-neuralmagic

**Comment by robertgshaw2-neuralmagic** [comment]
> I fully understand your comparison method and your results. Perhaps we should focus on how to improve the current AWQ's accuracy, what do you think? @robertgshaw2-neuralmagic @alexm-neuralmagic

Yes - the source of the drop in accuracy here seems to be the quality of the model, not the correctness of the marlin kernel. Neural Magic did not create this model and I do not know anything about how it was created. You can feel free to try to improve the accuracy as you see fit. Neural Magic provides quantized checkpoints on our Hugging Face profile that are compatible with vLLM with replication instructions for creation and evaluation of the models

The scope of this work is simply to run the model as fast as possible with accuracy scores the match the baseline implementations within numerical precision errors, which this PR accomplishes.

**Comment by mgoin** [comment]
@zhyncs I think you are confusing "AWQ, the quantization algorithm" versus "AWQ, the inference kernel". 
In vLLM we only have inference kernels (AWQ and AWQ Marlin are our current choices). These require models to be pre-quantized using the AWQ quantization algorithm, most notably [AutoAWQ](https://vllm--5388.org.readthedocs.build/en/5388/quantization/auto_awq.html) is the library used for that. 
Any inference engine that supports running AWQ models should give the same ~70% GSM8k accuracy reported in this thread, because this is a product of the quantized checkpoint (i.e. `casperhansen/llama-3-8b-instruct-awq`) rather than any inference code changing the model weights. This is why we say it isn't an issue of vLLM, but how the model was created (aka AutoAWQ with a specific config and calibration set). As Rob said, you said LMDeploy gives better AWQ results so please share what it gives when running `casperhansen/llama-3-8b-instruct-awq`.

If you want to reduce the impact of "AWQ, the quantization algorithm" you can produce your own quantized checkpoint using AutoAWQ with more conservative parameters. Looking at the [quantization config for that checkpoint](https://huggingface.co/casperhansen/llama-3-8b-instruct-awq/blob/main/config.json#L22), you can see it uses `"group_size": 128` which you could decrease to `"group_size": 64` when quantizing to improve accuracy preservation.

**Comment by casper-hansen** [comment]
This drop in accuracy after quantization looks normal to me. The standard calibration dataset used was not math-related either.

**Comment by robertgshaw2-neuralmagic** [comment]
Also, running the model through huggingface, I get the same scores as we get in vllm:

- launch
```bash
lm_eval --model hf --model_args pretrained=casperhansen/llama-3-8b-instruct-awq --tasks gsm8k --num_fewshot 8 --batch_size 16
```

- scores:
```bash
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.6967|±  |0.0127|
|     |       |strict-match    |     8|exact_match|↑  |0.6975|±  |0.0127|
```

**Comment by zhyncs** [comment]
> I think you are confusing "AWQ, the quantization algorithm" versus "AWQ, the inference kernel".

@mgoin You misunderstood, I was saying here that the precision of AWQ in LMDeploy is better, and did not say it ran the casperhansen/llama-3-8b-instruct-awq model. The quantization in LMDeploy is generated through `lmdeploy lite auto_awq`, using a group size of 128, without using anything from AutoAWQ. The gsm8k eval results previously measured did not show as much reduction in precision **percentage** as the test results here. You can verify it yourselves.

**Comment by zhyncs** [comment]
> Also, running the model through huggingface, I get the same scores as we get in vllm:
> 
> * launch
> 
> ```shell
> lm_eval --model hf --model_args pretrained=casperhansen/llama-3-8b-instruct-awq --tasks gsm8k --num_fewshot 8 --batch_size 16
> ```
> 
> * scores:
> 
> ```shell
> |Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
> |-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
> |gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.6967|±  |0.0127|
> |     |       |strict-match    |     8|exact_match|↑  |0.6975|±  |0.0127|
> ```

@robertgshaw2-neuralmagic Thank you for providing this reference.

**Comment by mgoin** [comment]
Okay @zhyncs thank you for clarifying that you are talking about a separate checkpoint than what you measured here in vLLM and that LMDeploy quantizes models itself. 
I think it is clear that we don't want to produce AWQ quantized models in vLLM because we would like to leave that up to pre-deployment libraries like AutoAWQ. Maybe for your issue here you should open an issue in AutoAWQ since you would like to see improved accuracy out of checkpoints produced from there.

**Comment by zhyncs** [comment]
> Also, running the model through huggingface, I get the same scores as we get in vllm:
> 
> * launch
> 
> ```shell
> lm_eval --model hf --model_args pretrained=casperhansen/llama-3-8b-instruct-awq --tasks gsm8k --num_fewshot 8 --batch_size 16
> ```
> 
> * scores:
> 
> ```shell
> |Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
> |-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
> |gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.6967|±  |0.0127|
> |     |       |strict-match    |     8|exact_match|↑  |0.6975|±  |0.0127|
> ```

Hi @robertgshaw2-neuralmagic After thinking about it, the eval results with hf for the AWQ checkpoint don't clarify anything since the auto_awq kernel is used during lm_eval as well. Therefore, in theory, there should be no difference from AWQ's original implementation in vLLM.

**Pull request merged**